### PR TITLE
Remove all json handling from rustc_serialize

### LINKED ISSUE #40177: Migrate rustc from rustc-serialize to serde
Cargo is doing it ( https://github.com/rust-lang/cargo/pull/3682 ). Crates.io is doing it ( https://github.com/rust-lang/crates.io/issues/583 ). We want them to think we're cool, so we need to do it too.

### PR DESCRIPTION
**Pull request by bjorn3**
Json is now handled using serde_json. Where appropriate I have replaced json usage with binary serialization (rmeta files) or manual string formatting (emcc linker arg generation).

This allowed for removing and simplifying a lot of code, which hopefully results in faster serialization/deserialization and faster compiles of rustc itself.

Where sensible we now use serde. Metadata and incr cache serialization keeps using a heavily modified (compared to crates.io) rustc-serialize version that in the future could probably be extended with zero-copy deserialization or other perf tricks that serde can't support due to supporting more than one serialization format.

Note that I had to remove `-Zast-json` and `-Zast-json-noexpand` as the relevant AST types don't implement `serde::Serialize`.

Fixes #40177

See also https://github.com/rust-lang/compiler-team/issues/418

### CODE REVIEW

**Comment by bjorn3** [comment]
@rust-timer build a0d810daa8d1883f9c5ce417226f96be2b28e2d7

**Comment by jyn514** [comment]
cc https://github.com/rust-lang/compiler-team/issues/418

**Comment by jyn514** [comment]
Apparently evaluate_obligation had a 20% regression after this (!!)

**Comment by bjorn3** [comment]
I reverted the changes to the `Encoder` and `Decoder` traits to see if they are the cause. My next guess would be 89b2df46df607df9dee82f18f413a1ba8bbba93e.

@bors try @rust-timer queue

**Comment by jyn514** [comment]
@bors r-

Perf looks much better this time :)

**Comment by bjorn3** [comment]
It looks better, but still a regression. Now trying if reverting the json diagnostics back to rustc-serialize improves perf.

@bors try @rust-timer queue

**Comment by tmiasko** [comment]
The earlier 31% regression appears to be there because `<rustc_middle::traits::ObligationCauseCode as core::hash::Hash>::hash` landed in wrong CGU, just like in an earlier PR that was merely moving code around https://perf.rust-lang.org/compare.html?start=066f01d81bfbed746f6b6cf27a0426d829e8e832&end=b3ac52646f7591a811fa9bf55995b24fd17ece08&stat=instructions:u

**Comment by jyn514** [comment]
@tmiasko Do you think it would help to mark that function as #[inline]?

**Comment by varkor** [comment]
I don't have much time to review large PRs at the moment. @jyn514: would you mind taking over the review?

**Comment by jyn514** [comment]
I don't have time either unfortunately :( I can leave comments here and there but I won't have time to review the whole diff.

**Comment by cjgillot** [comment]
TBH, I liked the first version of this PR better :smile: 
From @tmiasko's comment on diesel-doc, we should perhaps just discard this aberrant measurement. I find the globally greenish results and the massive simplification of the code worth it.

**Comment by varkor** [comment]
I'm going to reassign arbitrary because otherwise I'm just going to hold this up even longer.

r? @Aaron1011

**Comment by Aaron1011** [comment]
@bjorn3 Can you address the merge conflicts?

**Comment by Aaron1011** [comment]
The removal of `-Zast-json` and `-Zast-json-noexpand` needs a larger discussion - I suspect that this is going to make some nightly users very unhappy.

**Comment by JohnCSimon** [comment]
ping again from triage:
@bjorn3 Can you address the merge conflicts?

**Comment by bjorn3** [comment]
Rebased. Should I remove the revert commits? (https://github.com/rust-lang/rust/pull/85993#issuecomment-871755821)

**Comment by Aaron1011** [comment]
My earlier comment still stands - the removal of `-Zast-json` and `-Zast-json-noexpand` needs much broader discussion.

cc @rust-lang/compiler

**Comment by oli-obk** [comment]
@bjorn3 can we "just" implement the `serde::Serialize` trait for those ast types? Or is there a reason why we shouldn't?

**Comment by bjorn3** [comment]
That should be possible, but it would likely lead to code bloat, making compilation of rustc slower.

**Comment by oli-obk** [comment]
So... I think the fastest way to resolve this situation is to write an MCP. Though likely in the discussion people will ask if we know of any users of these flags, so looking for users may be a good first step.

**Comment by pnkfelix** [comment]
team discussed a bit. Agreed with @oli-obk : best way forward is for @bjorn3 (or someone else with an interest in this happening) to write an MCP for it.

**Comment by JohnCSimon** [comment]
Ping from triage: 
 @bjorn3 what is the status of this PR?

**Comment by nnethercote** [comment]
Note: #93681 removed the use of JSON for `.rlink` files.

**Comment by bjorn3** [comment]
Rebased again and updated commit messages.

@bors try @rust-timer queue

**Comment by Aaron1011** [comment]
It doesn't look like there was any mention of `-Zast-json` in the MCP https://github.com/rust-lang/compiler-team/issues/418. Before we remove it, we need to determine if there's a replacement for any users (including compiler developers) who are relying on it.

**Comment by bjorn3** [comment]
Yeah, I still need to open an MCP for that. The replacement would be using rustc_driver directly.

**Comment by nnethercote** [comment]
@bjorn3:  Instead of removing `Zast-json` and `-Zast-json-noexpand` in this PR, would it make sense to convert them to `serde_json`? Then all the other changes in this PR could land immediately, and the removal of those options could be done in a follow-up after doing an MCP.

I ask because I've been fiddling with the `Encoder` trait and it's a bit annoying having to make changes to `json::Encoder` and `json::PrettyEncoder` knowing that they're likely to be gone soon. So the sooner they're gone, the better. If this PR doesn't make progress soon I may well remove `json::Encoder` and change the one existing use point to instead use `json::PrettyEncoder`; that would at least halve the amount of JSON encoding code.

**Comment by bjorn3** [comment]
> Instead of removing Zast-json and -Zast-json-noexpand in this PR, would it make sense to convert them to serde_json?

I'm afraid that will cause quite some bloat. We would have to derive `serde::Serialize` in addition to `rustc_serialize::Encodable` as the later is still necessary for crate metadata encoding. I guess I can try it though.

https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-08-12.20.2354818/near/249241302

**Comment by nnethercote** [comment]
I see, that's unfortunate. Thanks for the explanation.

**Comment by pnkfelix** [comment]
https://github.com/rust-lang/compiler-team/issues/503 was accepted (:tada:), so if that was the only thing blocking progress here, its handled.

@rustbot label: -I-needs-decision

File: compiler/rustc_codegen_ssa/src/back/linker.rs [hunk 1]
```diff
@@ -1139,26 +1138,24 @@ impl<'a> Linker for EmLinker<'a> {
...
         self.cmd.arg("-s");
 
         let mut arg = OsString::from("EXPORTED_FUNCTIONS=");
-        let mut encoded = String::new();
-
-        {
-            let mut encoder = json::Encoder::new(&mut encoded);
-            let res = encoder.emit_seq(symbols.len(), |encoder| {
-                for (i, sym) in symbols.iter().enumerate() {
-                    encoder.emit_seq_elt(i, |encoder| encoder.emit_str(&("_".to_owned() + sym)))?;
```
Comment (Aaron1011) [inline compiler/rustc_codegen_ssa/src/back/linker.rs:1153]: It looks like the leading "_" is no longer included in your custom serialization code. Are there any tests verifying that the `EXPORTED_FUNCTIONS` we construct actually works?

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/linker.rs:1153]: Right, emscripten is a tier 2 platform so it doesn't get any testing.

File: src/test/ui-fulldeps/extern-mod-syntax.rs [hunk 2]
```diff
@@ -3,8 +3,8 @@
 #![allow(unused_imports)]
 #![feature(rustc_private)]
 
-extern crate rustc_serialize;
-use rustc_serialize::json::Object;
+extern crate libc;
+use libc::c_void;
```
Comment (bjorn3) [inline src/test/ui-fulldeps/extern-mod-syntax.rs:7]: Needed a replacement here. `c_void` is available in libc on all platforms I think. Not sure though.

File: src/test/ui-fulldeps/issue-11881.rs [hunk 3]
```diff
@@ -1,52 +0,0 @@
-// run-pass
```
Comment (bjorn3) [inline src/test/ui-fulldeps/issue-11881.rs:1]: Not sure what the exact original bug was so I don't know how to rewrite it to not depend on rustc_serialize.

File: src/test/ui-fulldeps/issue-15924.rs [hunk 4]
```diff
@@ -1,26 +0,0 @@
-// run-pass
```
Comment (bjorn3) [inline src/test/ui-fulldeps/issue-15924.rs:1]: Same

File: src/test/ui-fulldeps/issue-2804.rs [hunk 5]
```diff
@@ -1,70 +0,0 @@
-// run-pass
```
Comment (bjorn3) [inline src/test/ui-fulldeps/issue-2804.rs:1]: Same

File: compiler/rustc_codegen_ssa/src/back/linker.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/linker.rs:1160]: Fixed

**Comment by Aaron1011** [comment]
I'm a little hesitant about removing the methods like `emit_struct` - did they only ever exist for the purpose of encoding to JSON? Is there a possibility that we might want to use them again in the future as a way of optimizing metadata / incr-comp encoding?

**Comment by Aaron1011** [comment]
My thinking was that they could be used for some kind of automatic 'shorthand' encoding (like we do with certain types) that relies on knowing that several different pieces of data are associated together.

However, I don't think we'll ever need to add new target formats for `Encodable/Decodable`, so those methods should never be required for correctness. Unless anyone else objects, I guess we can just go ahead and remove them in this PR.

**Comment by bjorn3** [comment]
@Aaron1011 https://github.com/rust-lang/rust/pull/85993#discussion_r856298400 isn't resolved yet.

**Comment by Aaron1011** [comment]
I misread that comment, and thought you were intentionally removing that test as part of removing json support.

@bors r-

**Comment by nnethercote** [comment]
> I'm a little hesitant about removing the methods like `emit_struct` - did they only ever exist for the purpose of encoding to JSON? Is there a possibility that we might want to use them again in the future as a way of optimizing metadata / incr-comp encoding?

FWIW, the equivalent API simplifications have already been done for decoding, in #93839.

**Comment by bjorn3** [comment]
I think bors didn't notice the approval.

@bors r=Aaron1011

**Comment by bjorn3** [comment]
Got the manual json serialization for emscripten wrong. Switched it to serde_json, so it should be fine now.

**Comment by bjorn3** [comment]
Rebased

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
@bjorn3 feel free to r=me after rebasing. Thanks for your work on this!

**Comment by bjorn3** [comment]
Rebased

@bors r=wesleywiser

**Comment by nnethercote** [comment]
I was unable to reproduce the perf regressions in a local build. Some of the benchmarks in question (e.g. cranelift-codegen-0.82.1, keccak) are sensitive to small changes in codegen in and around `process_obligations`, I wonder if something happened there. Hopefully it's just noise and will shift back at some point?

**Pull request merged**